### DISCUSSION TITLE
Deprecate GCN3 and move exclusively to Vega / gfx9

### DISCUSSION THREAD

**Discussion by abmerop** [gem5/gem5#306] [gem5-dev]
I'd like to propose deprecating the GCN3 ISA and moving exclusively to Vega.  Vega is essentially a superset ISA of GCN3, meaning it contains everything GCN3 has and more.  As a result of this, there is an enormous amount of redundant code. gem5 also now supports an APU model with Vega as well as a dGPU so users can simulate either one, just like GCN3. 

The plan would be as follows:
- Update all GPU gem5-resources Makefiles to build the gfx902 target (Vega APU).  gfx900 (dGPU) is already being built.
- Ensure that all GPU gem5-resources tests work as well or better with Vega compared to GCN3.
- Update documentation for removal references to GCN3.
- Change the default --gfx-version to gfx902 or gfx900.
   - I think if we choose gfx900, we should also implicitly set the --dgpu flag as this is no gfx900 APU.
- Delete GCN3 related files.

The benefits would be:
- Removes 90k lines of code from the gem5 source code.
- Less applications to test and one less build target to test.
- A single GPU ISA means GPU can be built into build/ALL/gem5.* further reducing testing (I think - Ruby protocol might be an issue).
- gfx9 is the current ISA used in Radeon™ Instinct™ products up to and including the most recent hardware - So we would be simulating *current* products.

Will add more to plan/benefits if I think of anything else, but wanted to get the discussion started.

**Comment by mattsinc** [comment]
I agree with.  Two things to consider:

- We need to make sure all GPU applications we support in gem5-resources work in Vega
- For ML workloads (e.g., DNNMark), I'm pretty sure they are not supported in gfx902 -- I believe the MIOpen folks essentially do not support APUs (e.g., gfx902) and gfx902 only really works for applications that aren't using libraries (e.g., MIOpen).  I guess we could just acknowledge this in the documentation?

**Comment by abmerop** [comment]
Hi Matt,

I am glad you agree.  I haven't figured out DNNMark just yet and as part of the plan I am just getting started to test gfx902.  I am going to test if github will let me paste my status sheet, if so these are the results so far (as of 9/13/23): 

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/mporemba/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/mporemba/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



SE mode - develop |   | dgpu (gfx900) | apu (gfx902) | Notes
-- | -- | -- | -- | --
Heterosync | sleepMutex 10 16 4 | PASS |   |  
  | lfTreeBarrUniq 10 16 4 |   |   |  
HACC | 0.5 0.1 64 0.1 1 N 12 rcb | PASS |   |  
hsa-agent-pkt | N/A | N/A | N/A |  
lulesh | - | PASS | PASS |  
bc (pannotia) | 1k_128k.gr | FAIL |   | event: WrVicBlk state: M
color , maxmin   (pannotia) | 1k_128k.gr 0 | PASS |   |  
color, max (pannotia) | 1k_128k.gr 0 | PASS |   |  
fw (pannotia) | 1k_128k.gr 2 | PASS |   |  
mis (pannotia) | 1k_128k.gr 0 | PASS |   |  
pagerank, spmv   (pannotia) | coAuthorsDBLP.graph 1 | PASS |   |  
pagerank, default (pannotia) | coAuthorsDBLP.graph 1 | FAIL |   | event: WrVicBlk state: M   (Note:   somehow different than GPUFS)
sssp, ELL   (pannotia) | 1k_128k.gr 0 | PASS |   |  
sssp, CSR (pannotia) | 1k_128k.gr 0 | PASS |   |  
pennant | nohsmall | PASS |   |  
  | noh | FAIL |   | Hangs - Vega @ task 20, GCN3 @ task 3
square | - |   |   |  
DNNMark | -config   ${gem5_root}/gem5-resources/src/gpu/DNNMark/config_example/softmax_config.dnnmark |   |   | Can't find dnnmark.so
  | -config   ${gem5_root}/gem5-resources/src/gpu/DNNMark/config_example/pool_config.dnnmark
  | -config   ${gem5_root}/gem5-resources/src/gpu/DNNMark/config_example/bn_config.dnnmark |   |   |  



</body>

</html>

**Comment by abmerop** [comment]
Forgot the tl;dr:   Basically everything that works in GCN3 so far works just as well with Vega dGPU.

**Reply by mattsinc** [reply]
Looks like the HIP Example apps need to be tested too.  And yeah we'll need to figure out how to fix Pennant too.  And test gfx902 like you said.  Let me know how we can help.

**Comment by BobbyRBruce** [comment]
I'm very much on-board with this. It's a big thumbs-up for me.

I've got a simple (perhaps stupid) question on this: What does this mean for SE-mode simulation? Does this change anything? Will we still have the same docker image?

**Comment by abmerop** [comment]
Hi Bobby, not a stupid question and I am glad you agree!   In short: The docker will not change for this, we will just build different targets.

**Comment by abmerop** [comment]
Some updates: All SE mode tests pass (besides PENNANT noh which didn't work anyway).

website references updated here: https://github.com/gem5/website/pull/87
Working on gem5-resources but there is *a lot* of stuff to change and some unrelated things

**Reply by mattsinc** [reply]
Great!  Probably we should make PENNANT's smaller input the one that eventually goes in the new weekly test then?  Or make it the default recommended size on gem5-resources: https://gem5.googlesource.com/public/gem5-resources/+/refs/heads/develop/src/gpu/pennant/README.md?  I suppose we could update weekly.sh in the short-term if that helps...but probably not necessary.  I think PENNANT is not in weekly at all as is, so we'd just need to add it.

**Reply by abmerop** [reply]
I agree we should switch to nohsmall.  I was just updating the gem5-resources READMEs and thought I would do that.

The noh input is hanging (which you may already know).  I am hoping to apply my "gpu deadlock detector" changes to develop to see what is happening but have a few other things I want to get to first

**Reply by mattsinc** [reply]
sure, let me know when you need me to review (again).  I'm also fine waiting to see if you can find the source of the deadlock, or happy to dig through a trace with you too.  My go to for Ruby deadlocks is to print (e.g., in GPUCoalescer) when each request is sent/returned, then diff to see where the first sent one is not returned.  Not sure if that is relevant here?

**Reply by abmerop** [reply]
I want to test again with some of the recent changes, specifically with the waitcnt/tokens change.  I am not sure this is a Ruby deadlock but it could be related to tokens.  One way to check without the GPU deadlock detector I mentioned is to simply run for a while and force gem5 to dump stats (`kill -s SIGUSR1 <pid>`) and there will be a scoreboard stage stat that is 10-100x bigger than the other stall reasons which can give the user an idea

**Comment by BobbyRBruce** [comment]
There doesn't appear to be any disagreement with this idea. I've created issue #703 to track the progress of this task
